src/ring/middleware/json.clj

```diff
@@ -6,21 +6,23 @@
             [ring.core.protocols :as ring-protocols]
             [ring.util.io :as ring-io]
             [ring.util.response :refer [content-type]]
-            [ring.util.request :refer [character-encoding]]))
+            [ring.util.request :refer [character-encoding]])
+  (:import
+   [java.io InputStream]))
 
 (defn- json-request? [request]
   (if-let [type (get-in request [:headers "content-type"])]
     (not (empty? (re-find #"^application/(.+\+)?json" type)))))
 
 (defn- read-json [request & [{:keys [keywords? bigdecimals?]}]]
   (if (json-request? request)
-    (if-let [body (:body request)]
-      (let [encoding (or (character-encoding request)
+    (if-let [^InputStream body (:body request)]
+      (let [^String encoding (or (character-encoding request)
                           "UTF-8")
-            body-string (slurp body :encoding encoding)]
+            body-stream (java.io.InputStreamReader. body encoding)]
```

Comment:
`body-stream` should probably be `body-reader`.

Reply:
done